nifi-api/src/main/java/org/apache/nifi/documentation/AbstractDocumentationWriter.java

```diff
@@ -0,0 +1,301 @@
+ * a component.
+ *
+ * Please note that while this class lives within the nifi-api, it is provided primarily as a means for documentation components within
+ * the NiFi NAR Maven Plugin. Its home is the nifi-api, however, because the API is needed in order to extract the relevant information and
+ * the NAR Maven Plugin cannot have a direct dependency on nifi-api (doing so would cause a circular dependency). By having this homed within
+ * the nifi-api, the Maven plugin is able to discover the class dynamically and invoke the one or two methods necessary to create the documentation.
+ *
+ * This is a new capability in 1.9.0 in preparation for the Extension Registry and therefore, you should
+ * <b>NOTE WELL:</b> At this time, while this class is part of nifi-api, it is still evolving and may change in a non-backward-compatible manner or even be
+ * removed from one incremental release to the next. Use at your own risk!
+ */
+public abstract class AbstractDocumentationWriter implements DocumentationWriter {
+
+    @Override
+    public final void write(final ConfigurableComponent component) throws IOException {
+        write(component, null);
+    }
+
+    @Override
+    public final void write(final ConfigurableComponent component, final Collection<ProvidedServiceAPI> providedServices) throws IOException {
+        initialize(component);
+
+        writeHeader(component);
+        writeBody(component);
+
+        if (providedServices != null && component instanceof ControllerService) {
+            writeProvidedServices(providedServices);
+        }
+
+        writeFooter(component);
+    }
+
+    private void initialize(final ConfigurableComponent component) {
+        try {
+            if (component instanceof Processor) {
+                initialize((Processor) component);
+            } else if (component instanceof ControllerService) {
+                initialize((ControllerService) component);
+            } else if (component instanceof ReportingTask) {
+                initialize((ReportingTask) component);
+            }
+        } catch (final InitializationException ie) {
+            throw new RuntimeException("Failed to initialize " + component, ie);
+        }
+    }
+
+    protected void initialize(final Processor processor) {
+        processor.initialize(new DocumentationProcessorInitializationContext());
+    }
+
+    protected void initialize(final ControllerService service) throws InitializationException {
+        service.initialize(new DocumentationControllerServiceInitializationContext());
+    }
+
+    protected void initialize(final ReportingTask reportingTask) throws InitializationException {
+        reportingTask.initialize(new DocumentationReportingInitializationContext());
+    }
+
+    protected void writeBody(final ConfigurableComponent component) throws IOException {
+        final String additionalDetails = getAdditionalDetailsHtml(component);
+
+        writeExtensionName(component.getClass().getName());
```

Comment:
I think we would need "extensionType" or "extensionCategory" to indicate processor, cs, or reporting task, unless we wanted to use more specific elements instead of just <extension> </extension>

Reply:
I intended to include this, but apparently I missed it. Will update that.